pkg/operator/certrotation/target_test.go

```diff
@@ -631,3 +636,217 @@ func TestEnsureTargetSignerCertKeyPair(t *testing.T) {
+		Refresh:   12 * time.Hour,
+		CertCreator: &ClientRotation{
+			UserInfo: &user.DefaultInfo{Name: "system:kube-controller-manager"},
+		},
+
+		Client:        fakeKubeClient.CoreV1(),
+		Lister:        corev1listers.NewSecretLister(coreV1Indexer),
+		EventRecorder: events.NewInMemoryRecorder("test"),
+	}
+
+	// normally, we would have secondClientCertificate managed by secondCertController,
+	// but it is not needed to show the issue
+
+	// normally, firstCertController would call EnsureTargetCertKeyPair
+	firstClientCertResult, err := firstClientCertificate.EnsureTargetCertKeyPair(context.TODO(), firstSigner, []*x509.Certificate{firstSigner.Config.Certs[0]})
+	require.NoError(t, err)
+	coreV1Indexer.Add(firstClientCertResult)
+
+	// since the secondCertController won, eventually secondSignerSecret should be used
+	// to regenerate the firstClientCertificate but because the CommonName is exactly the same, but it isn't
+	firstClientCertSecondResult, err := firstClientCertificate.EnsureTargetCertKeyPair(context.TODO(), secondSigner, []*x509.Certificate{secondSigner.Config.Certs[0]})
+	require.NoError(t, err)
+
+	// the firstClientCertSecondResult should be signed with the secondSignerSecret
+	require.NotEqual(t, firstClientCertResult, firstClientCertSecondResult)
+}
+
+// TestCABundleWithMultipleSigners shows that a CABundle can miss a CAs
+func TestCABundleWithMultipleSigners(t *testing.T) {
+	firstSignerSecret := &corev1.Secret{}
+	require.NoError(t, setSigningCertKeyPairSecret(firstSignerSecret, time.Hour))
+	secondSignerSecret := &corev1.Secret{}
+	require.NoError(t, setSigningCertKeyPairSecret(secondSignerSecret, time.Hour))
+
+	firstSecretCommonName := firstSignerSecret.Annotations[CertificateIssuer]
+	secondSecretCommonName := secondSignerSecret.Annotations[CertificateIssuer]
+	require.Equal(t, firstSecretCommonName, secondSecretCommonName)
+
+	firstSigner, err := crypto.GetCAFromBytes(firstSignerSecret.Data["tls.crt"], firstSignerSecret.Data["tls.key"])
+	require.NoError(t, err)
+	secondSigner, err := crypto.GetCAFromBytes(secondSignerSecret.Data["tls.crt"], secondSignerSecret.Data["tls.key"])
+	require.NoError(t, err)
+
+	fakeKubeClient := kubefake.NewSimpleClientset()
+	coreV1Indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+
+	target := CABundleConfigMap{
+		Namespace:     "ns",
+		Name:          "trust-bundle",
+		Client:        fakeKubeClient.CoreV1(),
+		Lister:        corev1listers.NewConfigMapLister(coreV1Indexer),
+		EventRecorder: events.NewInMemoryRecorder("test"),
+	}
+
+	expectedCertificates := []*x509.Certificate{}
+
+	res, err := target.EnsureConfigMapCABundle(context.TODO(), firstSigner)
+	require.NoError(t, err)
+	_, err = fakeKubeClient.CoreV1().ConfigMaps("ns").Get(context.Background(), "trust-bundle", metav1.GetOptions{})
+	require.NoError(t, err)
+	// by not adding the caBundleConfigMap
+	// to the lister, we simulate the race condition
```

Comment:
This isn't correct. Not adding it to a lister will make sure it would be created from scratch, which is different from "racing update" as it doesn't have necessary metadata